lib/vexml/EasyScoreMessageProducer.ts

```diff
@@ -1,52 +1,71 @@
 import { MusicXMLCursor } from './MusicXMLCursor';
-import { MusicXMLParser } from './MusicXMLParser';
-import { EasyScoreMessage, EasyScoreMessageReceiver, XMLNode } from './types';
+import { EasyScoreMessage, EasyScoreMessageReceiver} from './types';
 
 export class EasyScoreMessageProducer {
-  static feed(musicXml: string) {
-    const root = MusicXMLParser.parse(musicXml);
-    return new EasyScoreMessageProducer(root);
+  static feed(musicXml: Document) {
+    return new EasyScoreMessageProducer(musicXml);
   }
 
-  private root: XMLNode;
+  private root: Document;
 
-  private constructor(root: XMLNode) {
+  private constructor(root: Document) {
     this.root = root;
   }
 
   message(receiver: EasyScoreMessageReceiver): void {
     const cursor = MusicXMLCursor.from(this.root);
+    receiver.onMessage({ type: 'voiceStart' });
     while (cursor.hasNext()) {
       cursor.next();
       const node = cursor.get();
-      const messages = this.getMessages(node);
+      console.log(node);
+      const messages: EasyScoreMessage[] = [];
+      messages.push(...this.getMessages(node));
       for (const message of messages) {
         receiver.onMessage(message);
       }
     }
+    receiver.onMessage({ type: 'voiceEnd' });
   }
 
-  private getMessages(node: XMLNode): EasyScoreMessage[] {
-    switch (node) {
-      case MusicXMLCursor.clefNode:
-        return [{ type: 'clef', clef: 'treble' }];
-      case MusicXMLCursor.timeSignatureNode:
-        return [{ type: 'timeSignature', top: 4, bottom: 4 }];
-      case MusicXMLCursor.measureNode:
-        return [
-          { type: 'voiceStart' },
-          { type: 'note', stem: 'up', pitch: 'C#5', duration: '1/4' },
-          { type: 'note', stem: 'up', pitch: 'B4', duration: '1/4' },
-          { type: 'note', stem: 'up', pitch: 'B4', duration: '1/8' },
-          { type: 'beamStart' },
-          { type: 'note', stem: 'up', pitch: 'F4', duration: '1/8' },
-          { type: 'note', stem: 'up', pitch: 'G4', duration: '1/8' },
-          { type: 'note', stem: 'up', pitch: 'G#4', duration: '1/8' },
-          { type: 'beamEnd' },
-          { type: 'voiceEnd' },
-        ];
+  private getMessages(node: Node): EasyScoreMessage[] {
```

Comment:
I like the `NoteHandler` idea, I wonder though if we need the intermediate `EasyScoreMessage` perhaps it would be better to use `Node` directy with some access functions to attributes and childs.

Reply:
The problem with using `Node` directly is that the `EasyScoreMessageReceiver` would become tightly coupled to the MusicXML structure. I would be fine with this if MusicXML was more compatible with `EasyScore`. However, as you saw with beams, you need some mechanism to "look ahead" to get the full context of the beam before you can tell `EasyScore` about it. There are so many other cases that would make this significantly complex.

Testing also becomes more difficult, which makes me think using `Node` -> `EasyScore` is not abstracted well enough. So instead of testing the following separately:

- `Node` -> `EasyScoreMessage`
- `EasyScoreMessage` -> `VexflowObjects`

You'd have to test: 

- `Node` -> `VexflowObjects`

All in all, I think the importance of `EasyScoreMessage` is that we can control the message API, and conform it to `EasyScore`. It allows someone else to implement other source formats, such as ABC notation -> `EasyScoreMessage` only having to do only part of the work.

I think we need to define state machines that describe the system we're trying to make. This imperative approach isn't going to scale well. Are you open to trying something like [xstate](https://xstate.js.org/)?